### PR TITLE
[Serve] Update dashboard to display gRPC proxy related metrics and info

### LINKED ISSUE #38741: [Serve] Improve gRPC proxy related metrics and dashboard
### Description

Some metrics in Grafana and Ray Dashboard are tracking HTTP specific metrics per routes. Let's add new panels for gRPC equivalent and tracking per gRPC methods.  

### Use case

_No response_

### PR DESCRIPTION
**Pull request by GeneDer**
## Why are these changes needed?

We deployed all the changes to allow users to use Ray Serve with gRPC proxy. However, the dashboard and metrics are yet to display gRPC related info. In this PR:
- Changed Serve Grafana dashboard to use per application instead of per route
- Added HTTP route and gRPC method as filters
- Exposed `grpc_options` to serve instance details api
- Modified dashboard frontend to accept and display gRPC port
- Updated default log message about HTTP proxy to also include gRPC proxy
- Update metrics doc for gRPC related changes

## Related issue number

Closes https://github.com/ray-project/ray/issues/38741

## Checks

- [ ] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
    - [ ] I've added any new APIs to the API Reference. For example, if I added a 
           method in Tune, I've added it in `doc/source/tune/api/` under the 
           corresponding `.rst` file.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

Rebuilt the dashboard frontend and restarted grafana to confirm the changes
image
image
image
image

### CODE REVIEW

File: python/ray/serve/tests/test_metrics.py [hunk 1]
```diff
@@ -15,8 +16,15 @@
...
 from ray.serve._private.common import DeploymentID
 from ray.serve._private.constants import DEFAULT_LATENCY_BUCKET_MS
+from ray.serve.config import gRPCOptions
 from ray.serve.drivers import DAGDriver
 from ray.serve.http_adapters import json_request
+from ray.serve.tests.test_grpc import (
+    ping_grpc_list_applications,
+    ping_grpc_call_method,
+    ping_fruit_stand,
+)
```
Comment (zcin) [inline python/ray/serve/tests/test_metrics.py:26]: Nit: these can be moved to `test/utils.py`.

File: python/ray/serve/tests/test_metrics.py [hunk 2]
```diff
@@ -274,13 +353,32 @@ def f(*args):
     assert num_deployment_errors[0]["application"] == "app"
     print("serve_num_deployment_http_error_requests working as expected.")
 
+    num_deployment_errors = get_metric_dictionaries(
+        "serve_num_deployment_grpc_error_requests"
+    )
+    assert len(num_deployment_errors) == 1
+    assert num_deployment_errors[0]["deployment"] == "app_f"
```
Comment (zcin) [inline python/ray/serve/tests/test_metrics.py:360]: You may have to merge master, the PR to remove application prefix was merged recently.

File: dashboard/modules/metrics/dashboards/serve_grafana_dashboard_base.json [hunk 3]
```diff
@@ -52,17 +52,17 @@
...
         "hide": 0,
         "includeAll": true,
         "label": null,
         "multi": true,
-        "name": "Route",
+        "name": "Application",
         "options": [],
         "query": {
-          "query": "label_values(ray_serve_num_http_requests{{{global_filters}}}, route)",
+          "query": "label_values({{__name__=~ 'ray_serve_num_http_requests|ray_serve_num_grpc_requests',{global_filters}}}, application)",
```
Comment (zcin) [inline dashboard/modules/metrics/dashboards/serve_grafana_dashboard_base.json:65]: Should we change this to `ray_serve_deployment_replica_healthy` (like what we use in `serve_deployment_grafana_dashboard_base.json`)? If no requests has been sent yet, might the application not show up in the dropdown?

**Review by zcin — COMMENTED** [review]
Great work, looks much cleaner to replace route with application! Could you also add to the serve agent test `test_get_serve_instance_details` to check on the grpc options?

File: python/ray/serve/tests/test_metrics.py (hunk 2 above)
Comment (GeneDer) [reply python/ray/serve/tests/test_metrics.py:360]: Yea, I was expecting to rebase again. At the time when I wrote this, that change is not in master, let me double check

File: dashboard/modules/metrics/dashboards/serve_grafana_dashboard_base.json (hunk 3 above)
Comment (GeneDer) [reply dashboard/modules/metrics/dashboards/serve_grafana_dashboard_base.json:65]: oh that's great call out! Let me update it

File: python/ray/serve/controller.py [hunk 4]
```diff
@@ -487,6 +488,12 @@ def get_http_config(self):
             return None
         return self.http_proxy_state_manager.get_config()
 
+    def get_grpc_config(self):
```
Comment (sihanwang41) [inline python/ray/serve/controller.py:491]: nit: add return type please.

File: python/ray/serve/controller.py [hunk 5]
```diff
@@ -487,6 +488,12 @@ def get_http_config(self):
...
+        """Return the gRPC proxy configuration."""
+        if self.http_proxy_state_manager is None:
```
Comment (sihanwang41) [inline python/ray/serve/controller.py:493]: as a follow up, we should rename the http proxy state manager.

File: python/ray/serve/controller.py [hunk 6]
```diff
@@ -885,10 +893,12 @@ def get_serve_instance_details(self) -> Dict:
         # route_prefix is set instead in each application.
         # Eventually we want to remove route_prefix from DeploymentSchema.
         http_options = HTTPOptionsSchema.parse_obj(http_config.dict(exclude_unset=True))
+        grpc_options = gRPCOptionsSchema.parse_obj(grpc_config.dict(exclude_unset=True))
```
Comment (sihanwang41) [inline python/ray/serve/controller.py:896]: do we have unit test for this change?

File: python/ray/serve/controller.py (hunk 5 above)
Comment (GeneDer) [reply python/ray/serve/controller.py:493]: Yes, I'm fully aware with will the name changes :)

**Comment by edoakes** [comment]
Two high-level suggestions:

1. Drop the gRPC method panes from Serve dashboard (if these breakdowns are needed, can click into Grafana).
2. Make both filter-by-route and filter-by-gRPC-method possible & opt-in in the grafana dashboard.

**Comment by sihanwang41** [comment]
For `Changed Serve Grafana dashboard to use per application instead of per route`,  what if we use fastapi and expose multiple routes inside the same application? Can you help to make sure it is not broken?

**Comment by GeneDer** [comment]
@sihanwang41 "pre route" part is currently broken due to the way we log those metrics unfortunately. Part of the advantage for this PR is also so we don't need to deal with the broken routes 😅

File: dashboard/modules/metrics/dashboards/serve_dashboard_panels.py [hunk 7]
```diff
@@ -51,42 +51,54 @@
...
+                expr='histogram_quantile(0.5, sum(rate(ray_serve_http_request_latency_ms_bucket{{application=~"$Application",application!~"",route=~"$HTTP_Route",route!~"/-/.*",{global_filters}}}[5m])) by (application, route, le))',
+                legend="{{application, route}}",
+            ),
+            Target(
+                expr='histogram_quantile(0.5, sum(rate(ray_serve_grpc_request_latency_ms_bucket{{application=~"$Application",application!~"",method=~"$gRPC_Method",{global_filters}}}[5m])) by (application, method, le))',
+                legend="{{application, method}}",
             ),
             Target(
-                expr='histogram_quantile(0.5, sum(rate(ray_serve_http_request_latency_ms_bucket{{route=~"$Route",route!~"/-/.*",{global_filters}}}[5m])) by (le))',
+                expr='histogram_quantile(0.5, sum(rate({{__name__=~ "ray_serve_http_request_latency_ms_bucket|ray_serve_grpc_request_latency_ms_bucket",application=~"$Application",application!~"",{global_filters}}}[5m])) by (le))',
```
Comment (alexeykudinkin) [inline dashboard/modules/metrics/dashboards/serve_dashboard_panels.py:101]: You can make this more succinct:

```
__name__=~ "ray_serve_(http|grpc)_request_latency_ms_bucket"
```

File: dashboard/modules/metrics/dashboards/serve_dashboard_panels.py [hunk 8]
```diff
@@ -96,16 +108,20 @@
...
+                expr='histogram_quantile(0.9, sum(rate(ray_serve_http_request_latency_ms_bucket{{application=~"$Application",application!~"",route=~"$HTTP_Route",route!~"/-/.*",{global_filters}}}[5m])) by (application, route, le))',
+                legend="{{application, route}}",
+            ),
+            Target(
+                expr='histogram_quantile(0.9, sum(rate(ray_serve_grpc_request_latency_ms_bucket{{application=~"$Application",application!~"",method=~"$gRPC_Method",{global_filters}}}[5m])) by (application, method, le))',
+                legend="{{application, method}}",
             ),
             Target(
-                expr='histogram_quantile(0.9, sum(rate(ray_serve_http_request_latency_ms_bucket{{route=~"$Route",route!~"/-/.*",{global_filters}}}[5m])) by (le))',
+                expr='histogram_quantile(0.9, sum(rate({{__name__=~ "ray_serve_http_request_latency_ms_bucket|ray_serve_grpc_request_latency_ms_bucket",application=~"$Application",application!~"",{global_filters}}}[5m])) by (le))',
```
Comment (alexeykudinkin) [inline dashboard/modules/metrics/dashboards/serve_dashboard_panels.py:124]: Please ditto everywhere

**Review by alexeykudinkin — APPROVED** [review]

**Review by edoakes — APPROVED** [review]
Awesome work.

File: dashboard/modules/metrics/dashboards/serve_dashboard_panels.py (hunk 7 above)
Comment (GeneDer) [reply dashboard/modules/metrics/dashboards/serve_dashboard_panels.py:101]: oh those are just regex, let me try it out, thanks for the suggestiong!

**Comment by edoakes** [comment]
@GeneDer pls address Alexey's comment and then assign Zhe w/ `tests-ok` once it looks good

File: python/ray/serve/_private/proxy_request_response.py [hunk 9]
```diff
@@ -139,7 +139,7 @@ def request_type(self) -> str:
 
     @property
     def method(self) -> str:
-        return "GRPC"
+        return self.method_name
```
Comment (alexeykudinkin) [inline python/ray/serve/_private/proxy_request_response.py:142]: @GeneDer can we actually include both _service_ and method name like following `service/method`

**Review by sihanwang41 — APPROVED** [review]

**Review by zcin — APPROVED** [review]

File: dashboard/modules/metrics/dashboards/serve_dashboard_panels.py [hunk 10]
```diff
@@ -51,42 +51,54 @@
...
+                expr='histogram_quantile(0.5, sum(rate(ray_serve_http_request_latency_ms_bucket{{application=~"$Application",application!~"",route=~"$HTTP_Route",route!~"/-/.*",{global_filters}}}[5m])) by (application, route, le))',
+                legend="{{application, route}}",
+            ),
+            Target(
+                expr='histogram_quantile(0.5, sum(rate(ray_serve_grpc_request_latency_ms_bucket{{application=~"$Application",application!~"",method=~"$gRPC_Method",{global_filters}}}[5m])) by (application, method, le))',
+                legend="{{application, method}}",
             ),
             Target(
-                expr='histogram_quantile(0.5, sum(rate(ray_serve_http_request_latency_ms_bucket{{route=~"$Route",route!~"/-/.*",{global_filters}}}[5m])) by (le))',
+                expr='histogram_quantile(0.5, sum(rate({{__name__=~ "ray_serve_(http|grpc)_request_latency_ms_bucket",application=~"$Application",application!~"",{global_filters}}}[5m])) by (le))',
```
Comment (alanwguo) [inline dashboard/modules/metrics/dashboards/serve_dashboard_panels.py:101]: I don't understand this promql syntax, what is this trying to do?

File: dashboard/modules/metrics/dashboards/serve_dashboard_panels.py [hunk 11]
```diff
@@ -96,16 +108,20 @@
...
+                expr='histogram_quantile(0.9, sum(rate(ray_serve_http_request_latency_ms_bucket{{application=~"$Application",application!~"",route=~"$HTTP_Route",route!~"/-/.*",{global_filters}}}[5m])) by (application, route, le))',
+                legend="{{application, route}}",
+            ),
+            Target(
+                expr='histogram_quantile(0.9, sum(rate(ray_serve_grpc_request_latency_ms_bucket{{application=~"$Application",application!~"",method=~"$gRPC_Method",{global_filters}}}[5m])) by (application, method, le))',
+                legend="{{application, method}}",
             ),
             Target(
-                expr='histogram_quantile(0.9, sum(rate(ray_serve_http_request_latency_ms_bucket{{route=~"$Route",route!~"/-/.*",{global_filters}}}[5m])) by (le))',
+                expr='histogram_quantile(0.9, sum(rate({{__name__=~ "ray_serve_(http|grpc)_request_latency_ms_bucket|ray_serve_grpc_request_latency_ms_bucket",application=~"$Application",application!~"",{global_filters}}}[5m])) by (le))',
```
Comment (alanwguo) [inline dashboard/modules/metrics/dashboards/serve_dashboard_panels.py:124]: same here

File: dashboard/modules/metrics/dashboards/serve_dashboard_panels.py (hunk 10 above)
Comment (GeneDer) [reply dashboard/modules/metrics/dashboards/serve_dashboard_panels.py:101]: In the old query it only looks for HTTP metrics. This change is to include also the gRPC data.

Comment (GeneDer) [reply dashboard/modules/metrics/dashboards/serve_dashboard_panels.py:101]: Also changed the filter to look for application which applies to both HTTP and gRPC.

**Comment by GeneDer** [comment]
@zhe-thoughts this one is ready to merge

**Review by zhe-thoughts — APPROVED** [review]
Yes this is a release blocker

**Pull request merged**